File: lib/aius_website/members/membership.ex
defmodule AiusWebsite.Members.Membership do
  use Ecto.Schema
  import Ecto.Changeset
  import AiusWebsite.Repo.Utils
  alias AiusWebsite.Term.Period
  alias AiusWebsite.Members.User

  schema "memberships" do
    field :valid, :boolean
    field :situation, :string

    belongs_to :user, User
    belongs_to :period, Period

    timestamps()
    authors()
  end

  @doc false
  def changeset(membership, attrs) do
    membership
    |> cast(attrs, [:valid, :situation, :user_id, :period_id])
    |> set_author(Map.get(attrs, "author"))
    |> assoc_constraint(:user)
    |> assoc_constraint(:period)
    |> validate_required([:valid, :situation])
  end
end

File: assets/src/models/membership.ts
import * as t from 'io-ts';

import Period, { PeriodV } from './period';

const MembershipV = t.type({
  id: t.number,
  valid: t.boolean,
  situation: t.string,
  period: PeriodV,
}, 'Membership');

interface Membership {
  id: number;
  valid: boolean;
  situation: string;
  period: Period;
};

export default Membership;
export { Membership, MembershipV };
